### PR TITLE
Import redecl chain of function template specializations

### PR DESCRIPTION
**Pull request by martong**
In case of function template specializations, their `DeclContext` was not set up correctly.
Also, the redecl chain was not set up. This caused an error since #329 where we just simply removed the related tests. For upstreaming any redecl chain related change, we had to re enable those tests and implement the redecl chain for specializations too.

### CODE REVIEW

File: unittests/AST/ASTImporterTest.cpp [hunk 1]
```diff
@@ -1501,6 +1501,313 @@ TEST_P(
...
+    auto FromD = LastDeclMatcher<FunctionDecl>().match(FromTU, Pattern);
+    Import(FromD, Lang_CXX);
+  }
+
+  Decl *ToTU = ImportedD->getTranslationUnitDecl();
+
+  EXPECT_EQ(DeclCounter<FunctionDecl>().match(ToTU, Pattern), 2u);
+  auto To0 = FirstDeclMatcher<FunctionDecl>().match(ToTU, Pattern);
+  auto To1 = LastDeclMatcher<FunctionDecl>().match(ToTU, Pattern);
+  EXPECT_TRUE(ImportedD == To0);
```
Comment (balazske) [inline unittests/AST/ASTImporterTest.cpp]: Maybe add: `Imported != To1` ?

Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]: Ok, updated the tests with that check where applicable.

File: lib/AST/ASTImporter.cpp [hunk 2]
```diff
@@ -2534,7 +2571,7 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
               D->hasExternalFormalLinkage()) {
             if (IsStructuralMatch(D, FoundFunction)) {
               if (D->doesThisDeclarationHaveABody() &&
-                  FoundFunction->doesThisDeclarationHaveABody())
+                  FoundFunction->hasBody())
```
Comment (balazske) [inline lib/AST/ASTImporter.cpp]: Why is this changed?

Comment (martong) [reply lib/AST/ASTImporter.cpp]: First I used `doesThisDeclarationHaveABody` in the case of function template specializations too.
But, there is a new test case (`ImportDefinitions`) which revealed an essential problem: `doesThisDeclarationHaveABody` is true only if that `FoundFunction` has a body, `hasBody` is true if there is one declaration with a body in the redecl chain of `FoundFunction`.
During parsing, `DeclContext::addDecl` will not add a new element to the StoredDeclsList rarther it simply overwrites the old element with the most recent one (see #386 ). And this most recent  decl will be found by lookup. However, the most recent decl can be a prototype without a body and in that case we will duplicate the `FunctionDecl`. Perhaps I could write a test for this case as well ...

Comment (balazske) [reply lib/AST/ASTImporter.cpp]: Is it possible that the `hasBody` will return true but the actual `FoundFunction` has no body (it exists in another Decl)? In this case a to-be imported FunctionDecl that has the body will be mapped to one that has no body (at that Decl). But I am not sure if this is problem.

**Comment by martong** [comment]
Ci job completed with no regression. Bitcoin 7 failures, protobuf 10 failures.
http://cc.elte.hu:15009/job/test-clang-with-projects-egbomrt/11/HTML_Report/
